### PR TITLE
Logging implementation

### PR DESCRIPTION
**Pull request by dalderliesten**
Added the basic logging system, and added support for game start/stop logging, game initialization file handling and termination handling, player position and player input logging, and shooting logging.

### CODE REVIEW

**Comment by jmeijer8** [comment]
It is good implemented.  I see you worked yourself the blubber on this assignment.  Goodwork, thanks for taking this plan upon you.

File: src/main/java/nl/tudelft/scrumbledore/Logger.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    try {
+      // Fetching the current date for the creation of the file and parsing it to allow for simple
+      // formatting.
+      Date currentDate = new Date();
+      SimpleDateFormat simpleFormat = new SimpleDateFormat("yyyy-MM.dd-hh-mm-ss");
+
+      // Taking care of the file name and file creation.
+      String desiredFileName = "Session-" + simpleFormat.format(currentDate) + ".log";
+      loggingFile = new File(Constants.RESOURCES_DIR + Constants.LOGGER_DIR + desiredFileName);
```
Comment (jessetilro) [inline src/main/java/nl/tudelft/scrumbledore/Logger.java]: Couldn't the code above this line be outside the try statement's block?

File: src/main/java/nl/tudelft/scrumbledore/GUI.java [hunk 2]
```diff
@@ -423,10 +441,22 @@ public void handle(KeyEvent keyPressed) {
 
         // Mapping the desired keys to the desired actions.
         if (keyPress.equals("LEFT")) {
+          // Logging the movement of the player to the game log.
+          Logger.log("Player moved to " + player.getPosition().getX() + ", "
+              + player.getPosition().getY());
```
Comment (jessetilro) [inline src/main/java/nl/tudelft/scrumbledore/GUI.java]: The keypress event handler doesn't tell much about the movement of the player. For example the Player could be standing next to a wall, and therefore the keypress would not result in any movement. Even if it did, it's coordinates would not be updated instantaniously, that takes a step. So wouldn't it be better to call logs like this for example "User pressed LEFT key" instead?

File: src/main/java/nl/tudelft/scrumbledore/Scrumbledore.java [hunk 3]
```diff
@@ -17,6 +17,9 @@
    *          Arguments given at program initialization
    */
   public static void main(String[] args) {
+    // Create a logger to log all actions in this session.
+    Logger.start();
```
Comment (jessetilro) [inline src/main/java/nl/tudelft/scrumbledore/Scrumbledore.java]: Neat

File: src/main/java/nl/tudelft/scrumbledore/GUI.java [hunk 4]
```diff
@@ -366,6 +372,9 @@ private void advanceLevel() {
    */
   private void checkPlayerAlive() {
     if (!game.getCurrentLevel().getPlayer().isAlive()) {
+      // Logging that the game has been restarted.
+      Logger.log("--------------------PLAYER RESTARTED THE GAME--------------------");
```
Comment (jessetilro) [inline src/main/java/nl/tudelft/scrumbledore/GUI.java]: Maybe add a function to the logger adding these "-" signs automatically in the right amount so every log line has the same length?

[feature]

Comment (FlorisXll) [inline src/main/java/nl/tudelft/scrumbledore/GUI.java]: Why not just leave the "-" signs and let the sentence start at the beginning of the line?

Comment (jessetilro) [inline src/main/java/nl/tudelft/scrumbledore/GUI.java]: For the first version, I think that might be best indeed.

File: src/main/java/nl/tudelft/scrumbledore/Logger.java (hunk 1 above)
Comment (NielsWarnars) [inline src/main/java/nl/tudelft/scrumbledore/Logger.java]: As far as I know one always has to catch or throw an exception when handling files like what is the case here.

Exceptions that could occur might for example be caused by a file which is already opened in another program or which does not exist, etc.

File: src/main/java/nl/tudelft/scrumbledore/GUI.java (hunk 4 above)
Comment (dalderliesten) [inline src/main/java/nl/tudelft/scrumbledore/GUI.java]: The reason there are two events with 20 lines surrounding the log is that these are events that control the game flow.  For example, restarting the game, pausing the game, terminating the game, etc.  Useful for future implementation of, for example, a replay mode.

File: src/main/java/nl/tudelft/scrumbledore/Logger.java (hunk 1 above)
Comment (jessetilro) [inline src/main/java/nl/tudelft/scrumbledore/Logger.java]: Of course, I agree, but the code mentioned above does not throw such exceptions. Instantiation of the Date and the File class does not check the conditions you are mentioning, they don't throw an IOException.

Comment (dalderliesten) [inline src/main/java/nl/tudelft/scrumbledore/Logger.java]: All actions related to the file creation, including naming, must be within the try-catch statement for correct error handling.  Niels explained why, in terms of concurrent file accessing.

File: src/main/java/nl/tudelft/scrumbledore/GUI.java (hunk 2 above)
Comment (dalderliesten) [inline src/main/java/nl/tudelft/scrumbledore/GUI.java]: Sure thing, where would you encourage I do movement tracking?  Directly from player,pos?

Comment (jessetilro) [inline src/main/java/nl/tudelft/scrumbledore/GUI.java]: Maybe we should first discuss what exactly we want to log and when we want to log it. Logging a Player's position whenever it is updated might result in a lot of unnecessary lines in the log.

File: src/main/java/nl/tudelft/scrumbledore/Logger.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  /**
+   * Allows the caller to log the passed string in the current sessions' loging file.
+   * 
+   * @param toLog
+   *          The content that the caller wishes to be logged in the logging file.
+   */
+  public static void log(String toLog) {
+    try {
```
Comment (jessetilro) [inline src/main/java/nl/tudelft/scrumbledore/Logger.java]: The start method repeats this code, wouldn't it be possible to call the log method from within the start method? (Like log("SCRUMBLEDORE LOGGIN FILE");)

Comment (dalderliesten) [inline src/main/java/nl/tudelft/scrumbledore/Logger.java]: It's possible, but it's not proper style considering I've decided as per the earlier comments in this pull request to do all handling of instantiation in the initial method.

<organization

File: src/main/java/nl/tudelft/scrumbledore/GUI.java (hunk 2 above)
Comment (dalderliesten) [inline src/main/java/nl/tudelft/scrumbledore/GUI.java]: Could you check the latest commit?

**Comment by jessetilro** [comment]
Nice, I think this basic for the Logger implementation is ready to be merged, approved.

**Pull request merged**